src/pip/_internal/req/req_uninstall.py

```diff
@@ -183,6 +183,93 @@ def compress_for_output_listing(paths):
     return will_remove, will_skip
 
 
+class StashedUninstallPathSet(object):
+    """A set of file rename operations to stash files while
+    tentatively uninstalling them."""
+    def __init__(self):
+        self._save_dirs = {}
+        self.moves = []
+
+    def add_root(self, path):
+        """Adds a root directory that we will be moving files from."""
+        if not os.path.isdir(path):
+            raise ValueError("Roots must be directories")
+
+        # Keep return values internal
+        self._add_root(path)
```

Comment:
For readability, I think it would help if (generally speaking and where possible) methods could be defined before their first use (so `_add_root()` before `add_root()` in this case). That way the class can be read from top to bottom, with lower-level methods first.

Reply:
Sure, though personally I prefer it the other way around for the same reason (higher-level logic before lower-level logic)